contracts/casino/include/casino/casino.hpp

```diff
@@ -97,102 +102,111 @@ class [[eosio::contract("casino")]] casino: public eosio::contract {
+        game_state.modify(itr, get_self(), [&](auto& row) {
+            row.balance += quantity;
+        });
+        gstate.game_profits_sum += quantity;
     }
 
     void sub_balance(uint64_t game_id, asset quantity) {
-        game_state_table game_state(_self, _self.value);
-        auto itr = game_state.find(game_id);
-        if (itr == game_state.end()) {
-            game_state.emplace(get_self(), [&](auto& row) {
-                row.game_id = game_id;
-                row.quantity = -quantity;
-                row.last_claim_time = current_time_point();
-            });
-        } else {
-            game_state.modify(itr, get_self(), [&](auto& row) {
-                row.quantity -= quantity;
-            });
-        }
-        gstate.game_profits_sum -= quantity * get_profit_margin(game_id) / 100;
+        const auto itr = game_state.require_find(game_id, "game not found");
+        game_state.modify(itr, get_self(), [&](auto& row) {
+            row.balance -= quantity;
+        });
+        gstate.game_profits_sum -= quantity;
     }
 
     bool is_active_game(uint64_t game_id) const {
-        game_table games(_self, _self.value);
         return games.find(game_id) != games.end();
     }
 
     time_point get_last_claim_time(uint64_t game_id) const {
-        game_state_table game_state(_self, _self.value);
-        const auto itr = game_state.find(game_id);
-        return itr == game_state.end() ? time_point() : itr->last_claim_time;
+        const auto itr = game_state.require_find(game_id, "game not found");
+        return itr->last_claim_time;
     }
 
     void update_last_claim_time(uint64_t game_id) {
-        game_state_table game_state(_self, _self.value);
-        const auto itr = game_state.require_find(game_id, "game is not in the casino");
+        const auto itr = game_state.require_find(game_id, "game is not listed in the casino");
         game_state.modify(itr, _self, [&](auto& row) {
             row.last_claim_time = current_time_point();
         });
     }
 
-    void transfer(name to, asset quantity) {
+    void transfer(name to, asset quantity, std::string memo) {
         eosio::action(
             eosio::permission_level{_self, "active"_n},
             "eosio.token"_n,
             "transfer"_n,
-            std::make_tuple(_self, to, quantity, std::string("player winnings"))
+            std::make_tuple(_self, to, quantity, memo)
         ).send();
     }
 
     void verify_game(uint64_t game_id);
     void verify_account(name game_account);
     uint64_t get_game_id(name game_account);
 
-    void session_update(asset quantity) {
+    void session_update(uint64_t game_id, asset quantity) {
+        const auto itr = game_state.require_find(game_id, "game not found");
+        game_state.modify(itr, _self, [&](auto& row) {
+            row.active_sessions_sum += quantity;
+        });
         gstate.game_active_sessions_sum += quantity;
```

Comment:
`gstate.game_active_sessions_sum` and  `row.active_sessions_sum` invariant is greater that zero, so we need to check that in any places where we change that values.

Reply:
`gstate.game_active_sessions_sum` and `row.active_sessions_sum` could be zero

Reply:
ok, >=